plugin/evm/orderbook/trading_apis.go

```diff
@@ -344,8 +344,22 @@ func (api *TradingAPI) PlaceOrder(order *hu.SignedOrder) (common.Hash, error) {
 		return orderId, hu.ErrNoTradingAuthority
 	}
 
-	fields := api.db.GetOrderValidationFields(orderId, trader, marketId)
-	// @todo P1 - P3
+	fields := api.db.GetOrderValidationFields(orderId, order)
+	// P1. Order is not already in memdb
+	if fields.Exists {
+		return orderId, hu.ErrOrderAlreadyExists
+	}
+
+	if !order.ReduceOnly {
+		// P2. Margin is available for non-reduce only orders
+		requiredMargin := hu.Div1e18(hu.Mul(order.BaseAssetQuantity, order.Price))
```

Comment:
shouldn’t it be `required = price * quantity / 5`  ?

Reply:
Good catch. Fixed.